### ISSUE TITLE
[Enhancement]: Optimize expression parsing efficiency

### ISSUE THREAD

**Issue by xiaocai2333** [milvus-io/milvus#36672]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What would you like to be added?

Optimize expression parsing efficiency.
1. Optimize the time of parsing Chinese characters.
2. Pass in the expression parameters to be queried through the template.
  - such as  `expr=pk > {age} and city in {city}, search_params={age:3, city:['beijing","shanghai"]}`

**Comment by xiaofan-luan** [comment]
Excited about the template feature!

**Comment by zhengbuqian** [comment]
we now support nullable fields, how do we handle null in template?

**Comment by xiaocai2333** [comment]
> we now support nullable fields, how do we handle null in template?

Currently, our expression does not support null values. We can discuss it further when null support is added in the future.
/cc @smellthemoon

**Comment by xiaofan-luan** [comment]
null support is already added?

**Comment by xiaofan-luan** [comment]
did you mean isNull and isNotNull?
can we do a == null?

**Comment by zhengbuqian** [comment]
> did you mean isNull and isNotNull?

yes. we should be able to filter based on `null`. chatted with @smellthemoon offline, we don't currently support this.

Maybe we should not allow `null` in template, we can simply use a `NULL` keyword:

* to use null only: `a == NULL` or `a != NULL`
* to use null with non null values: `a == 100 || a == NULL`
* to use null with template: `a in [{accepted_values}] || a == NULL`

**Comment by smellthemoon** [comment]
> > did you mean isNull and isNotNull?
> 
> yes. we should be able to filter based on `null`. chatted with @smellthemoon offline, we don't currently support this.
> 
> Maybe we should not allow `null` in template, we can simply use a `NULL` keyword:
> 
> * to use null only: `a == NULL` or `a != NULL`
> * to use null with non null values: `a == 100 || a == NULL`
> * to use null with template: `a in [{accepted_values}] || a == NULL`

In fact, in the future, we will still not support expressions like == null and != null, which will directly report errors at the parse layer. Exposing the NULL keyword will bring a lot of complexity. It needs to be identified in all expressions, and the respective logical relationships need to be considered in complex expressions. Therefore, in the future, we will only support the form of isNull(field) and isNotNull(field).

**Comment by zhengbuqian** [comment]
> > > did you mean isNull and isNotNull?
> > 
> > 
> > yes. we should be able to filter based on `null`. chatted with @smellthemoon offline, we don't currently support this.
> > Maybe we should not allow `null` in template, we can simply use a `NULL` keyword:
> > 
> > * to use null only: `a == NULL` or `a != NULL`
> > * to use null with non null values: `a == 100 || a == NULL`
> > * to use null with template: `a in [{accepted_values}] || a == NULL`
> 
> In fact, in the future, we will still not support expressions like == null and != null, which will directly report errors at the parse layer. Exposing the NULL keyword will bring a lot of complexity. It needs to be identified in all expressions, and the respective logical relationships need to be considered in complex expressions. Therefore, in the future, we will only support the form of isNull(field) and isNotNull(field).

that is fine. as long as we have a way of checking null, we can remove the need to support null in expression template.
